File: crates/brainwires-mcp-server/src/middleware/tool_filter.rs
use async_trait::async_trait;
use brainwires_mcp_client::{JsonRpcError, JsonRpcRequest};
use std::collections::HashSet;

use super::{Middleware, MiddlewareResult};
use crate::connection::RequestContext;

/// Tool filtering mode.
pub enum FilterMode {
    /// Only allow listed tools.
    AllowList(HashSet<String>),
    /// Deny listed tools, allow all others.
    DenyList(HashSet<String>),
}

/// Middleware that filters tool calls by allow/deny lists.
pub struct ToolFilterMiddleware {
    mode: FilterMode,
}

impl ToolFilterMiddleware {
    /// Create a filter that only allows the specified tools.
    pub fn allow_only(tools: impl IntoIterator<Item = impl Into<String>>) -> Self {
        Self {
            mode: FilterMode::AllowList(tools.into_iter().map(|t| t.into()).collect()),
        }
    }

    /// Create a filter that denies the specified tools.
    pub fn deny(tools: impl IntoIterator<Item = impl Into<String>>) -> Self {
        Self {
            mode: FilterMode::DenyList(tools.into_iter().map(|t| t.into()).collect()),
        }
    }

    fn is_allowed(&self, tool_name: &str) -> bool {
        match &self.mode {
            FilterMode::AllowList(allowed) => allowed.contains(tool_name),
            FilterMode::DenyList(denied) => !denied.contains(tool_name),
        }
    }
}

#[async_trait]
impl Middleware for ToolFilterMiddleware {
    async fn process_request(
        &self,
        request: &JsonRpcRequest,
        _ctx: &mut RequestContext,
    ) -> MiddlewareResult {
        // Only filter tools/call
        if request.method != "tools/call" {
            return MiddlewareResult::Continue;
        }

        let tool_name = request
            .params
            .as_ref()
            .and_then(|p| p.get("name"))
            .and_then(|n| n.as_str())
            .unwrap_or("unknown");

        if self.is_allowed(tool_name) {
            MiddlewareResult::Continue
        } else {
            MiddlewareResult::Reject(JsonRpcError {
                code: -32001,
                message: format!("Tool '{tool_name}' is not allowed by filter policy"),
                data: None,
            })
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_allow_list() {
        let filter = ToolFilterMiddleware::allow_only(["agent_spawn", "agent_list"]);
        assert!(filter.is_allowed("agent_spawn"));
        assert!(filter.is_allowed("agent_list"));
        assert!(!filter.is_allowed("bash"));
    }

    #[test]
    fn test_deny_list() {
        let filter = ToolFilterMiddleware::deny(["bash", "write_file"]);
        assert!(!filter.is_allowed("bash"));
        assert!(!filter.is_allowed("write_file"));
        assert!(filter.is_allowed("agent_spawn"));
    }
}

File: deno/packages/mcp-server/middleware/tool_filter.ts
/**
 * @module middleware/tool_filter
 *
 * Tool filtering middleware (allowlist/blocklist).
 * Equivalent to Rust's `ToolFilterMiddleware`.
 */

import type { JsonRpcRequest } from "@rullama/mcp-client";
import type { RequestContext } from "../server.ts";
import {
  type Middleware,
  middlewareContinue,
  middlewareReject,
  type MiddlewareResult,
} from "./mod.ts";

/** Tool filtering mode. */
export type FilterMode =
  | { type: "allowList"; tools: Set<string> }
  | { type: "denyList"; tools: Set<string> };

/**
 * Middleware that filters tool calls by allow/deny lists.
 * Only applies to tools/call requests.
 * Equivalent to Rust `ToolFilterMiddleware`.
 */
export class ToolFilterMiddleware implements Middleware {
  private readonly mode: FilterMode;

  private constructor(mode: FilterMode) {
    this.mode = mode;
  }

  /** Create a filter that only allows the specified tools. */
  static allowOnly(tools: string[]): ToolFilterMiddleware {
    return new ToolFilterMiddleware({
      type: "allowList",
      tools: new Set(tools),
    });
  }

  /** Create a filter that denies the specified tools. */
  static deny(tools: string[]): ToolFilterMiddleware {
    return new ToolFilterMiddleware({
      type: "denyList",
      tools: new Set(tools),
    });
  }

  /** Check if a tool name is allowed by the current filter. */
  isAllowed(toolName: string): boolean {
    if (this.mode.type === "allowList") {
      return this.mode.tools.has(toolName);
    }
    return !this.mode.tools.has(toolName);
  }

  // deno-lint-ignore require-await
  async processRequest(
    request: JsonRpcRequest,
    _ctx: RequestContext,
  ): Promise<MiddlewareResult> {
    // Only filter tools/call
    if (request.method !== "tools/call") {
      return middlewareContinue();
    }

    const params = request.params as Record<string, unknown> | undefined;
    const toolName = (params?.name as string | undefined) ?? "unknown";

    if (this.isAllowed(toolName)) {
      return middlewareContinue();
    }

    return middlewareReject({
      code: -32001,
      message: `Tool '${toolName}' is not allowed by filter policy`,
    });
  }
}
